text/0000-bind-helper.md

```diff
@@ -0,0 +1,124 @@
+- Start Date: 2019-03-20
+- Relevant Team(s): Ember.js, Learning
+- RFC PR: https://github.com/emberjs/rfcs/pull/470
+- Tracking: (leave this empty)
+
+# Bind Helper
+
+## Summary
+
+This RFC introduces a new helper `bind` to allow clear argument and context scope binding for functions passed in templates.
+
+## Motivation
+
+The current `action` helper has a lot of implicit and confusing behavior that is different than the Octane and post Octane programming model.
+
+To understand the complexity of `action` there are many complex behaviors including:
+
+1. Argument partial application (currying)
+2. `this` context binding
+3. `send` checks for Component and Controllers
+
+At build time the `action` helper currently is passed through an AST transform to explicitly bind the `this` to be deterministic at runtime. This is a private API where the outputted Glimmer is not a 1-1 to the template. Also, the `action` helper is confused and has overlap with the `action` modifier which has similar but slightly different behavior.
+
+Instead of this confusing and overloaded behavior, a new `bind` helper would be introduced to do partial application and context binding (with no need for build time private APIs).
+
+## Detailed design
+
+The `bind` helper will take in a function and then the set of arguments that will be partially applied to the function.
+As an optional named argument the user can pass in a `this` property that will set the `this` context this named argument would default to `undefined`.
+
+The behavior of this helper will be simple and use JavaScript's `function.prototype.bind`.
+
+Here are some examples of the `bind` helper and the equivalent JS:
+
+### Simple Case On Argument Curry
+
+```hbs
+{{bind this.log 1}}
+```
+
+```js
+this.log.bind(null, 1);
+```
+
+### Multiple Argument Partial Application
+
+```hbs
+{{bind this.add 1 2}}
+```
+
+```js
+this.add.bind(null, 1, 2);
+```
+
+### Setting the `this` context
+
+```hbs
+{{{bind this.session.logout context=this.anotherSession}}}
+```
+
+```js
+this.session.logout.bind(this.session)
+```
+
+### Setting the `this` context and arguments
```

Comment:
imo this is a little confusing since its the opposite of what the JavaScript bind function argument order is (wrt the context)

Can we make use of the [array helper](https://github.com/emberjs/rfcs/pull/318) instead? Similar to how Angle Bracket does positional params (https://guides.emberjs.com/release/reference/syntax-conversion-guide/#toc_params-array)

```hbs
{{bind this.car.drive context=this.honda params=(array 1000 'mile')}}
```

Reply:
Doing this would be making the use of this helper too verbose and a step backwards with respect to the `action` helper.

Reply:
Yeah, we debated about this a bit, it comes down to “naming is hard”. What we want is a partial application helper. We’ve discussed:

* `partial` - used by aome other languages, but we can’t because it means something in hbs
* `apply` - also used by other languages, but already used in JS, and means something very different
* `curry` - could work, but then we have to explain the term. Also, may not be _technically_ correct (currying produces a function that recieves exactly 1 argument, or something like that)
* `papply` - easy to misinterpet if you don’t know what the term “partial application”
* `partial-apply` - fairly verbose for a common task

`bind` is the _least_ bad choice IMO, it’s like if you made a `bind` function with a slightly better API, but I also agree it’s not ideal and could be confusing. I think `curry` is the second best out of those options, and am open to suggestions!

Reply:
but isnt is confusing considered that in 90% of the cases `{{bind` will *not* be used to bind context but `@action` will bind context?

Reply:
I got confused by that too, @luxferresum. If `@action` will be the recommended way to bind context, I prefer a helper called `curry`, `apply-args` or something on that direction to partially apply arguments.